src/recon_buildblock/PLSPrior.cxx

```diff
@@ -0,0 +1,607 @@
+                                                                       inner_product[z][y][x]*inner_product[z][y][x]);
+                    }
+                 }
+          }
+
+}
+
+template <typename elemT>
+double
+PLSPrior<elemT>::
+compute_value(const DiscretisedDensity<3,elemT> &current_image_estimate)
+{
+  if (this->penalisation_factor==0)
+  {
+    return 0.;
+  }
+
+
+  shared_ptr<DiscretisedDensity<3,elemT> > pet_im_grad_z_sptr(this->anatomical_sptr->get_empty_copy ());
+  shared_ptr<DiscretisedDensity<3,elemT> > pet_im_grad_y_sptr(this->anatomical_sptr->get_empty_copy ());
+  shared_ptr<DiscretisedDensity<3,elemT> > pet_im_grad_x_sptr(this->anatomical_sptr->get_empty_copy ());
+
+  shared_ptr<DiscretisedDensity<3,elemT> > inner_product_sptr(this->anatomical_sptr.get ()->get_empty_copy ());
+  shared_ptr<DiscretisedDensity<3,elemT> > penalty_sptr(this->anatomical_sptr.get ()->get_empty_copy ());
+
+  compute_inner_product_and_penalty (*inner_product_sptr,
+                                     *penalty_sptr,
+                                     *pet_im_grad_z_sptr,
+                                     *pet_im_grad_y_sptr,
+                                     *pet_im_grad_x_sptr,
+                                     current_image_estimate);
+
+  const bool do_kappa = !is_null_ptr(kappa_ptr);
+
+  if (do_kappa && !kappa_ptr->has_same_characteristics(current_image_estimate))
+    error("PLSPrior: kappa image has not the same index range as the reconstructed image\n");
+
+
+  double result = 0.;
+  const int min_z = current_image_estimate.get_min_index();
+  const int max_z = current_image_estimate.get_max_index();
+  for (int z=min_z; z<=max_z; z++)
+    {
+
+      const int min_y = current_image_estimate[z].get_min_index();
+      const int max_y = current_image_estimate[z].get_max_index();
+
+        for (int y=min_y;y<= max_y;y++)
+          {
+
+            const int min_x = current_image_estimate[z][y].get_min_index();
+            const int max_x = current_image_estimate[z][y].get_max_index();
+
+            for (int x=min_x;x<= max_x;x++)
+              {
+
+                /* formula:
+                  sum_x,y,z
+
+                   (penalty[z][y][x]) * (*kappa_ptr)[z][y][x];
+                */
+
+                        elemT current = (*penalty_sptr)[z][y][x];
+
+                        if (do_kappa)
+                          current *=
+                            (*kappa_ptr)[z][y][x] ;
+
+                        result += static_cast<double>(current);
+
+              }
+          }
+    }
+  return result * this->penalisation_factor;
+}
+
+template <typename elemT>
+void
+PLSPrior<elemT>::
+compute_gradient(DiscretisedDensity<3,elemT>& prior_gradient,
+                 const DiscretisedDensity<3,elemT> &current_image_estimate)
+{
+  assert(  prior_gradient.has_same_characteristics(current_image_estimate));
+  if (this->penalisation_factor==0)
+  {
+    prior_gradient.fill(0);
+    return;
+  }
+
```

Comment:
needs a check
```c++
if (!this->anatomical_sptr->has_same_characteristics(current_image_estimate))
  error("blabla");
```
take the opportunity to convert the `assert` above to a real check as well